programs/util.c

```diff
@@ -174,6 +174,228 @@ U64 UTIL_getTotalFileSize(const char* const * const fileNamesTable, unsigned nbF
+
+  FILE* inputFile = fopen(inputFileName, "r");
+  int nbFiles = -1;
+  unsigned pos = 0;
+
+
+  if(!buf) {
+    UTIL_DISPLAYLEVEL(1, "[ERROR][UTIL_readFileNamesTableFromFile] Can't create buffer.\n");
+    return -1;
+  }
+
+  if(!inputFile) {
+    UTIL_DISPLAYLEVEL(1, "[ERROR][UTIL_readFileNamesTableFromFile] Can't open file to read input file names.\n");
+    return -1;
+  }
+
+  for(nbFiles=0; !feof(inputFile) ; ) {
+    if(UTIL_readLineFromFile(buf+pos, inputFileSize, inputFile) > 0) {
+      int len = (int) strlen(buf+pos);
+      buf[pos+len-1] = '\0'; /* replace '\n' with '\0'*/
+      pos += len;
+      ++nbFiles;
+    }
+  }
+
+  fclose(inputFile);
+
+  if(pos > inputFileSize) return -1;
+
+  return nbFiles;
+}
+
+/*Note: buf is not freed in case function successfully created table because filesTable->fileNames[0] = buf*/
+FileNamesTable*
+UTIL_createFileNamesTable_fromFileName(const char* inputFileName) {
+    U64 inputFileSize = 0;
+    unsigned nbFiles = 0;
+    int ret_nbFiles = -1;
+    char* buf = NULL;
+    size_t i = 0, pos = 0;
+
+    FileNamesTable* filesTable = NULL;
+
+    if(!UTIL_fileExist(inputFileName) || !UTIL_isRegularFile(inputFileName))
+      return NULL;
+
+    inputFileSize = UTIL_getFileSize(inputFileName) + 1; /* (+1) to add '\0' at the end of last filename */
+
+    if(inputFileSize > MAX_FILE_OF_FILE_NAMES_SIZE)
+      return NULL;
+
+    buf = (char*) malloc((size_t) inputFileSize * sizeof(char));
+    if(!buf) {
+      UTIL_DISPLAYLEVEL(1, "[ERROR][UTIL_readFileNamesTableFromFile] Can't create buffer.\n");
+      return NULL;
+    }
+
+    ret_nbFiles = readFromFile(buf, (size_t) inputFileSize, inputFileName);
+
+    if(ret_nbFiles <= 0) {
+      free(buf);
+      return NULL;
+    }
+    nbFiles = ret_nbFiles;
+
+    filesTable = UTIL_createFileNamesTable(NULL, NULL, 0);
+    if(!filesTable) {
+      free(buf);
+      UTIL_DISPLAYLEVEL(1, "[ERROR][UTIL_readFileNamesTableFromFile] Can't create table for files.\n");
+      return NULL;
+    }
+
+    filesTable->tableSize = nbFiles;
+    filesTable->fileNames = (const char**) malloc((nbFiles+1) * sizeof(char*));
+
+
+
+    for(i = 0, pos = 0; i < nbFiles; ++i) {
+      filesTable->fileNames[i] = buf+pos;
+      pos += strlen(buf+pos)+1;
+    }
+
+
+    if(pos > inputFileSize){
+      UTIL_freeFileNamesTable(filesTable);
+      if(buf) free(buf);
+      return NULL;
+    }
+
+    filesTable->buf = buf;
+
+    return filesTable;
+}
+
+FileNamesTable*
+UTIL_createFileNamesTable(const char** filenames, char* buf, size_t tableSize){
+  FileNamesTable* table = (FileNamesTable*) malloc(sizeof(FileNamesTable));
+  if(!table) {
+    return NULL;
+  }
+  table->fileNames = filenames;
+  table->buf = buf;
+  table->tableSize = tableSize;
+  return table;
+}
+
+void UTIL_freeFileNamesTable(FileNamesTable* table) {
+  if(table) {
+    if(table->fileNames) {
+      free(table->fileNames);
```

Comment:
It seems to be the last test error : 
`free(table->fileNames);` emits some warning on Visual, though it seems to be a Visual bug.

`table->fileNames` is a `const char**`, aka a table of `const char*`.
It's not possible to free a `const char*`, because operations on `const char*` shall not change any related state.
But it should be possible to `free()` a mutable table of pointers, even if those pointers point to some read-only memory. In this respect `const char* *table` should be no different from `object *table`.

However, Visual seems to be confused by the first `const`, and emits a warning. I think this is wrong. I would classify it as a bug, though it has been so long lasting, that I'm not even sure if it's a bug of a feature.

Anyway, the usual fix for this case is to cast the pointer to `(void*)`.